From 82ad865c3d72ad62b3d2d859c7501f4d956c2c37 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ond=C5=99ej=20Sur=C3=BD?= <ondrej@sury.org>
Date: Sat, 21 Jan 2017 07:58:46 +0100
Subject: [PATCH] When NOTIMPL or FORMERR is encountered, disable QNAME
 minimization first, and disable EDNS in a second round.  This adds some
 latency (one more roundtrip), but it's safer to not drop EDNS0 on a first
 failure.

---
 lib/layer/iterate.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/lib/layer/iterate.c b/lib/layer/iterate.c
index e50a3cf..ea3fca5 100644
--- a/lib/layer/iterate.c
+++ b/lib/layer/iterate.c
@@ -725,12 +725,16 @@ static int prepare_query(kr_layer_t *ctx, knot_pkt_t *pkt)
 
 static int resolve_badmsg(knot_pkt_t *pkt, struct kr_request *req, struct kr_query *query)
 {
+
 #ifndef STRICT_MODE
 	/* Work around broken auths/load balancers */
 	if (query->flags & QUERY_SAFEMODE) {
 		return resolve_error(pkt, req);
+	} else if (query->flags & QUERY_NO_MINIMIZE) {
+		query->flags | QUERY_SAFEMODE;
+		return KR_STATE_DONE;
 	} else {
-		query->flags |= QUERY_SAFEMODE;
+		query->flags |= QUERY_NO_MINIMIZE;
 		return KR_STATE_DONE;
 	}
 #else
-- 
2.11.0

